### PR TITLE
Unicode escapes are now ordinary escape sequences (not processed early)

### LINKED ISSUE #3220: Scalac should remove comments before translating unicode escapes
Consider this code:
```scala
object LinePrinter {
  def main(args: Array[String]){
    // Note: \u000A is Unicode representation of linefeed (LF)
    val c: Char = 0x000A
    print(c)
  }
}
```
This code does not compile, giving the following error method:

```scala
LinePrinter.scala:3: error: not found: value is
  // Note: \u000A is Unicode representation of linefeed (LF)
                  ^
one error found
```

The problem is that the compiler transforms unicode escapes before removing the comments, leading to the following code given to the next compile phase:

```scala
object LinePrinter {
  def main(args: Array[String]){
    // Note:
 is Unicode representation of linefeed (LF)
    val c: Char = 0x000A
    print(c)
  }
}
```

This feature/bug is described in the Java book "Java Puzzlers" (see the sampler, it is Puzzle 2) as "thoroughly confusing". Corresponding Java spec is [JLS 3.4].

I would like to suggest removing lines of comments before converting escapes, because how it works now is highly surprising and unexpected.

### PR DESCRIPTION
**Pull request by martijnhoekstra**
Unicode escapes are handled like any other escape, in any context a normal escape could also work.

That means Unicode escapes get expanded in

* Char literals
* String literals
* s and f interpolations
* back quoted identifiers

Examples:

```scala
'\u0061' == 'a'
"\u0061a" == "aa"
s"\u0061$a" == "aa"
f"\u0061$a" == "aa"
val `snowy_\u2603\u2603` = 17
snowy_☃☃ == 17
```

You can now use a Unicode escape sequence to represent a single quote in a single quoted s or f interpolation, in other words, s"\u0022foo\u0022" is now a valid way to represent `"\"foo\""`

Triple quoted strings and raw interpolations don't allow normal escapes, but have allowed unicode escapes, since those were handled earlier, in the scanner. In this PR doing so emits a deprecation warning, unless under -xSource:2.14, where the Unicode escapes are simply unprocessed

Examples:

```scala
"""triple \u0061""" == "triple a" // emits deprecation warning
raw"raw \u0061" == "raw a" // emits deprecation warning
```

Using Unicode escapes anywhere else will no longer cause them to be processed. In all those cases the alternative is to use the literal character represented by the escape sequence.

This includes interpolations. If your custom interpolation needs to process unicode escapes, it should do so from within the interpolator rather than relying on the parser to process these escapes.

example

```scala
mycustominterpolator"not \u0061n escape"
//interpolator will see the literal sequence \u0061
```

Examples that no longer compile (with no deprecation):

```scala
val \u0061 = 'a'
va\u0072 x = 7
```

Fixes https://github.com/scala/bug/issues/3220

(Curious why it ever worked the old way? Because Java did it, as explained at https://www.sitepoint.com/java-unicode-mysterious-compile-error/)

### CODE REVIEW

**Comment by martijnhoekstra** [comment]
Regarding `-X-no-uescape`, this PR gives it the axe (-X?) in the spirit of https://github.com/scala/scala-dev/issues/430

The setting was passed to `JavaByteArrayReader`, but not escaping Unicode escapes in java source was never valid in the first place per jls 3.3.

Removing that parameter outright is both source and binary breaking though, and probably not worth the deprecation cycle (though I'm happy to accommodate whatever if greater minds think unlike)

**Comment by martijnhoekstra** [comment]
Linking https://github.com/scala/bug/issues/3220 which this closes.

@SethTisue could you kick off a community build for this branch?

**Comment by SethTisue** [comment]
testing 2.13.1-bin-19737c1-SNAPSHOT: https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2432/

**Comment by SethTisue** [comment]
```
[scala-parser-combinators] [error] scala.StringContext$InvalidEscapeException: invalid escape '\[' not one of [\b, \t, \n, \f, \r, \, \", \', \uxxxx] at index 37 in "([^"\x00-\x1F\x7F\\]|\\[\\'"bfnrt]|\\u[a-fA-F0-9]{4})*". Use \\ for literal \.
```

```
[fast-string-interpolator] [info] - should don't compile in case of escaping error *** FAILED *** (11 milliseconds)
[fast-string-interpolator] [info]   "Expected no compiler error, but got the following type error: "invalid escape '\d' not one of [\b, \t, \n, \f, \r, \, \", \', \uxxxx] at index 0 in "\d". Use \\ for literal \.", for code:  fs"\d" " did not contain "invalid escape '\d' not one of [\b, \t, \n, \f, \r, \\, \", \'] at index 0 in "\d". Use \\ for literal \." (FastStringInterpolatorSpec.scala:34)
```

**Comment by martijnhoekstra** [comment]
@SethTisue could you that give another go with 9b61a1b? The fast-string-interpolator error is expected, as it checks the wording of a changed exception message, but parser-combinators is hopefully fixed now

**Comment by martijnhoekstra** [comment]
@SethTisue could you start another community build? The parser combinators failures are hopefully fixed now (and the other one was spurious)

**Comment by SethTisue** [comment]
queued: https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2439/

**Comment by martijnhoekstra** [comment]
I'm seeing a lot of community build stuff going on, but I can't figure out where, if anywhere, this branch is running now. Could you take a look and squeeze one in if you're not too swamped (or the servers) with the build itself?

**Comment by SethTisue** [comment]
oops, I shouldn't have canceled 2439, I got confused as well. (Jenkins doesn't let you attach a description to a run until the run actually starts)

queued: https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/2449/

**Comment by SethTisue** [comment]
consider that a green run — the handful of failures are unrelated to your changes

**Comment by martijnhoekstra** [comment]
No longer WIP. Not sure whom to request for review, there is a bit of spec, a bit of parser, a bit of interpolation.

**Comment by volth** [comment]
So, `s"\u0022$a\u0022"` is now valid ?
After deprecation `s"\042$a\042"` in 2.13 it is great news.

**Comment by martijnhoekstra** [comment]
It should. It's a good test case to add.

**Comment by martijnhoekstra** [comment]
Rebased for conflicts on mima exceptions.

@volth see https://github.com/scala/scala/pull/8282/commits/a83a51719355770866dec153a8920e46e26f8044 for a tes of that behaviour -- it works.

**Comment by martijnhoekstra** [comment]
rebased again for mima exceptions.

**Comment by szeiger** [comment]
What's the state of this in Dotty? I'd almost call this an uncontroversial bug fix that could probably be done without a SIP but we should at least have Dotty on board with this change. Rescheduling for 2.13.2 for now.

**Comment by volth** [comment]
> Rescheduling for 2.13.2 for now.

`2.12.next` would be nice too

**Comment by martijnhoekstra** [comment]
I didn't port this to dotty, so I would assume there is no state in dotty ATM.

I don't intend on backporting to 2.12 myself, but I suspect it would be welcomed if this is accepted.

**Comment by SethTisue** [comment]
I suggest we leave 2.12.x alone. 2.12 is in maintenance mode, and this is a design change and not an implementation fix

**Comment by martijnhoekstra** [comment]
@smarter are you able to answer the question of @szeiger?

**Comment by smarter** [comment]
I assume it does whatever scalac was doing so far, I don't personally have the time/energy to port this to Dotty.

**Comment by martijnhoekstra** [comment]
Is porting this to dotty required?

**Comment by smarter** [comment]
No such requirements have been established for scalac PRs yet, but if it doesn't get ported to Dotty now, there's a very real chance that it will be forgotten about and won't be part of 3.0.

**Comment by martijnhoekstra** [comment]
I don't think I'll be able to port this to Dotty.

**Comment by smarter** [comment]
FWIW, the Dotty scanner is pretty close to the scalac one, so I don't think a port of the scanner changes would be too hard. The changes to StringContext in the standard library would require a new release of scala-library for Dotty to depend on first, so that's a good argument for getting this PR merged first.

**Comment by smarter** [comment]
Also, instead of a dotty PR, an issue opened against dotty to make sure we remember to eventually port this change would be fine too.

**Comment by SethTisue** [comment]
@martijnhoekstra perhaps someone reading this like would like to volunteer to port the change to Dotty? seems like a good getting-started project for someone interested in becoming a Dotty contributor

**Comment by martijnhoekstra** [comment]
I don't really know, it doesn't sound too hard, though the crash on
incremental compilation makes it difficult for me personally. The
coordination with the StringContext changes may be tricky, I don't know how
that would work exactly.

Perhaps it's easier to wait for 3.0, and then target 3.0.1

**Comment by smarter** [comment]
What crash on incremental compilation ?

**Comment by martijnhoekstra** [comment]
The one in sbt: https://github.com/lampepfl/dotty/issues/7135

**Comment by smarter** [comment]
Ah right, I think the only solution is to switch to WSL, or give up on Windows, or rewrite sbt's classloader handling from scratch.

> Perhaps it's easier to wait for 3.0, and then target 3.0.1

I'm not sure this is a change that can go into a minor release.

**Comment by martijnhoekstra** [comment]
On Sun, Sep 8, 2019, 17:42 Guillaume Martres <notifications@github.com>
wrote:

> Ah right, I think the only solution is to switch to WSL, or give up on
> Windows, or rewrite sbt's classloader handling from scratch.
>
> Perhaps it's easier to wait for 3.0, and then target 3.0.1
>
> I'm not sure this is a change that can go into a minor release.
>

3.1.0 is always an option too.

**Comment by lrytz** [comment]
@martijnhoekstra I re-opened this PR, I think it should be merged for 2.13.2 (once it's reviewed, and 2.13.1 is released). The question how we manage forward-porting changes from Scala 2 to 3 is orthogonal to the improvements implemented here. The fact that we haven't figured it out yet doesn't block improvements such as this one from happening in Scala 2. We will also not condition contributions to Scala 2 on having a forward-port to Scala 3 available.

**Comment by martijnhoekstra** [comment]
That's a nice surprise.

While trying to port to dotty, I came across some issues that I want to double-check here (does it not accidentally interpret an unicode escape as terminator).

File: src/compiler/scala/tools/reflect/FastStringInterpolator.scala [hunk 1]
```diff
@@ -29,18 +29,29 @@ trait FastStringInterpolator extends FormatInterpolator {
...
+          parts.mapConserve { case lit@Literal(Constant(stringVal: String)) =>
+            val k = Constant(if (isRaw && currentRun.isScala214) stringVal
+                             else if (isRaw) {
+                               val processed = StringContext.processUnicode(stringVal)
+                               if(processed != stringVal){
+                                 val diffindex = processed.zip(stringVal).zipWithIndex.collectFirst {
+                                   case ((p, o), i) if p != o => i
+                                 }.getOrElse(processed.length - 1)
+                                 
+                                 c.warning(lit.pos.withShift(diffindex), "Unicode escapes in raw interpolations are deprecated as of scala 2.13.1, and will be removed in scala 2.14. Use the literal character instead.")
```
Comment (lrytz) [inline src/compiler/scala/tools/reflect/FastStringInterpolator.scala]: You can use `currentRun.reporting.deprecationWarning`

File: test/files/neg/t3220-1.check [hunk 2]
```diff
@@ -0,0 +1,25 @@
+t3220-1.scala:2: error: invalid unicode escape
+  val badsingle = "foo \unope that's wrong"
+                         ^
+t3220-1.scala:3: error: invalid unicode escape at index 6 of foo \unope that's wrong
+  val badtriple = """foo \unope that's wrong"""
+                        ^
```
Comment (lrytz) [inline test/files/neg/t3220-1.check]: caret off?

Comment (martijnhoekstra) [reply test/files/neg/t3220-1.check]: pesky off-by-3 errors

**Comment by martijnhoekstra** [comment]
How can I coordinate this with dotty? This touches both the library (for escaping in s interpolations) and the compiler. Making the same change in dotty depends on the updated library.

**Comment by smarter** [comment]
We upgrade the scala-library we depend on every time there's a release, so once this is part of a release the compiler changes can be done in dotty.

**Comment by martijnhoekstra** [comment]
the failing tests were the semantic conflict in the changes in error reporting that popped up when I rebased this: `x errors` vs `x errors found`.

**Comment by SethTisue** [comment]
I've put this back on the 2.13.2 milestone, because it seems nearly (entirely?) ready for merge, and based on Lukas's feedback so far.

(it will need squashing before final merge, of course)

**Comment by lrytz** [comment]
I support merging this for 2.13.2. To summarize the breaking changes:

  - Interpolations outside strings, char literals and backquoted identifiers are no longer supported. `val \u0061 = 'a'` will no longer compile, there's no deprecation.
  - Some breaking changes in corner cases because unicode escapes are expanded at a later stage. For example, `"\u0022"` used to be `"""`, now it's `"\""`. This enables a new workaround for https://github.com/scala/bug/issues/6476.

There's currently no migration tool (?) - would it be hard to write a scalafix rule? Can we include a one-liner in the release notes that fixes arbitrary code bases?

**Comment by martijnhoekstra** [comment]
There is a third and somewhat insidious (source, but not binary) breaking change: If you have a custom interpolator, unicode is no longer escaped in the interpolator.

I don't know about a scalafix. Looking at https://astexplorer.net/#/gist/ec56167ffafb20cbd8d68f24a37043a9/a34662ddd426356d811d29103e557e608c0e9f14 it is plausible it's not so easy to detect with scalameta.

**Comment by SethTisue** [comment]
as I said at the team meeting today, I'm comfortable with this change for 2.13.2 despite the string interpolator thing, because:

* we're epoch-major-minor not major-minor-point. a change like this is totally fair game IMO
* grepping a codebase for unicode escapes is easy, we can put a command in the release notes
* I strongly suspect most codebases have few or no such escapes, and manually reviewing them is little burden, so a scalafix seems like a would-maybe-be-nice to me rather than something we should feel obligated to provide
* this is already a biggish release with a lot of potentially breaking changes, it's not like including this will make the difference between "no-brainer upgrade" and "upgrade, but be careful", it's already the latter

**Comment by lrytz** [comment]
🎉 epochal, https://github.com/scala/bug/issues/3220 was submitted on Mar 27, 2010.

**Comment by dwijnand** [comment]
> * Interpolations outside strings, char literals and backquoted identifiers are no longer supported. `val \u0061 = 'a'` will no longer compile, there's no deprecation.

Just catching up, thank god sbt is 2.12 only 😄  https://github.com/sbt/sbt/blob/0b12862caf63d1c23449391be6329d947b474157/main-settings/src/main/scala/sbt/std/InputWrapper.scala#L26-L31

**Comment by martijnhoekstra** [comment]
The lines you link wouldn't be a problem (they're in strings), but the methods themselves should then be literal or backquoted in this PR -- either

    def `wrapTask_\u2603\u2603` =

or 

    def wrapTask_☃☃ = 


That wouldn't be that terrible a migration now, right?

**Comment by smarter** [comment]
> We upgrade the scala-library we depend on every time there's a release, so once this is part of a release the compiler changes can be done in dotty.

BTW, you could also already open a PR against dotty that uses a nightly build of scala-library, so we can review it in time for 2.13.2.

**Comment by Jasper-M** [comment]
> we're epoch-major-minor not major-minor-point.

Strictly speaking, isn't Scala epoch-major-point?

**Comment by SethTisue** [comment]
> Strictly speaking, isn't Scala epoch-major-point?

Good question!

If you mean the standard library and the ABI and nothing else, then yes.

But considering the releases as wholes, no. We often add features in minor releases. So that already makes them "minor" and not "point".

Admittedly, it is rarer that we break backwards source compatibility, as we are doing here. It's not something we should be cavalier about, for sure, but I've already listed some mitigating factors, and I'll add one more: there is no longer any 2.14 release to defer these sorts of changes to, so I think that argues for opening the door a little wider on allowable 2.13.x changes. This series will be with us for a long time.

(If this is turning into a broader discussion about the 2.13.x release philosophy, we should probably move to contributors.scala-lang.org. But if there are further comments that still relate directly to this change in particular, that's fine here.)

**Comment by SethTisue** [comment]
@martijnhoekstra can I ask you to revise the PR description so it contains all of the need-to-know information for users in one place? in particular, the info on breaking changes that came out in later comments

**Comment by martijnhoekstra** [comment]
Sure!

**Pull request merged**